bugwarrior/docs/services/logseq.rst

```diff
@@ -0,0 +1,150 @@
+-----------------------
+
+Logseq application must active with the HTTP API server running for bugwarrior to connect.
+
+To use bugwarrior to pull tickets from Logseq you need to enable the Logseq HTTP APIs server.
+In Logseq go to **Settings** > **Features** and toggle the **HTTP APIs server** option.
+
+Next select the **API** option in the top menu to configure authorization token, e.g
+
+.. image:: pictures/logseq_token.png
+
+
+Example Service
+---------------
+
+Here's an example of a Logseq target:
+
+.. config::
+
+    [my_logseq_graph]
+    service = logseq
+    logseq.host = localhost
+    logseq.port = 12315
+    logseq.token = mybugwarrioraccesstoken
+
+The above example is the minimum required to import issues from Logseq.
+You can also feel free to use any of the configuration options described in
+:ref:`common_configuration_options` or described in `Service Features`_ below.
+
+Service Features
+----------------
+
+Task filters
+++++++++++++
+
+You can configure the service to import tasks in different states.
+By default the service will import all tasks in an active tasks states
+
+    DOING, TODO, NOW, LATER, IN-PROGRESS, WAIT, WAITING
+
+You can override this filter by setting the ``logseq.task_state`` option to a 
+comma separated list of required task states.
+
+.. config::
+    :fragment: logseq
+
+    logseq.task_state = DOING, NOW, IN-PROGRESS
+
+Priority mapping
+++++++++++++++++
+
+Logseq task priorities ``A``, ``B``, and ``C`` mapped to the taskwarrior priorities
+``H``, ``M``, and ``L`` respectively.
+
+Task state and data/time mappings
++++++++++++++++++++++++++++++++++
+
+``DOING``, ``TODO``, ``NOW``, ``IN-PROGRESS`` are mapped to the default ``pending`` state.
+The Logseq task ``SCHEDULED:`` and ``DEADLINE:`` fields are mapped to the ``scheduled`` and 
+``due`` date fields.
+
+``LATER``, ``WAITING``, ``WAIT`` are mapped to the ``waiting`` state.
+The ``SHEDULED:`` date or ``DEADLINE`` date is used to set the ``wait`` date on the task.
+If no scheduled or deadlines date is available then the wait date is set to ``someday`` 
+(see ``Date and Time Synonyms <https://taskwarrior.org/docs/dates/#synonyms-hahahugoshortcode30s0hbhb/>``_).
+Waiting tasks can be listed using ``task waiting``
+
+``DONE`` is mapped to the ``completed`` state.
+
+``CANCELED`` and ``CANCELLED`` are mapped to the ``deleted`` state.
+
+Character replacement
++++++++++++++++++++++
+
+taskwarrior encodes the ``[`` and ``]`` characters commonly used in Logseq as ``&open;`` and ``&close;``. To
+avoid display issues ``[[`` and ``]]`` are replaced by ``【`` and ``】`` for page links, and single
+``[`` and ``]`` are replaced by ``〈`` and ``〉``. 
+
+You can override behaviour and use customer characters by setting the ``logseq.char_*`` options in your
+``bugwarriorrc`` config.
+
+.. config::
+    :fragment: logseq
+
+    logseq.char_open_link = 〖
+    logseq.char_close_link = 〗
+    logseq.char_open_bracket = (
+    logseq.char_close_bracket = )
```

Comment:
I think [this comment](https://github.com/GothenburgBitFactory/bugwarrior/pull/1052#discussion_r1650043792) belongs here:

> This is a work around for what is possibly an underlying bug in taskwarrior. The `[` and `]` characters in the task description, and other fields, are replaced with `&open;` and `&close;`.

Hmm, at a cursory glance it seems like Taskwarrior is correctly encoding them but failing to decode in `src/Task.cpp`:

```cpp
////////////////////////////////////////////////////////////////////////////////
// Encode values prior to serialization.
//   [  -> &open;
//   ]  -> &close;
const std::string Task::encode (const std::string& value) const
{
  auto modified = str_replace (value,    "[", "&open;");
  return          str_replace (modified, "]", "&close;");
}

////////////////////////////////////////////////////////////////////////////////
// Decode values after parse.
//   [  <- &open;
//   ]  <- &close;
const std::string Task::decode (const std::string& value) const
{
  if (value.find ('&') == std::string::npos)
    return value;

  auto modified = str_replace (value,    "&open;",  "[");
  return          str_replace (modified, "&close;", "]");
}
```

@scross01 What version of taskwarrior are you on?

@djmitche Any thoughts on this potential Taskwarrior bug?

Reply:
I'm using taskwarrior 3.0.2 on MacOS, installed using Homebrew

Reply:
I cannot reproduce the decoding bug on the command line by adding a task with brackets in the description on either tw-2 or tw-3. I'm not sure why bugwarrior would be any different since it is using the command line interface under the hood. @scross01 Can you verify manually adding and reading a task on the command line with brackets is or is not buggy on your system? Also, are you seeing the bug with the default report (e.g., running `task 42`) or are you using something fancier?

Reply:
This has been a long-standing bug, which I thought I reported once upon a time but can't find in https://github.com/GothenburgBitFactory/taskwarrior/issues right now. I filed https://github.com/GothenburgBitFactory/taskwarrior/issues/3524 just now.

Reply:
I ran
```shell
task add "test [[Task]]"
```
and running `task list` returns the result with
```
22 1min                                         test [[Task]]                                           0
```

I removed the escape code from the integration, with a task in Logseq called 
```
TODO Test Logseq [[Task]]
```
then ran `bugwarrior pull` (note was previously synced with the escaping in place)
```
INFO:bugwarrior.db:Updating task a633286d-c62c-4ec5-88c0-669ff29a2e3a, (bw)Is#126 - TODO Test Logseq [[Task]]; logseqid: '126' -> 126; logseqtitle: 'TODO Test Logseq 〖Task〗' -> 'TODO Test Logseq [[Task]]'; description: '(bw)Is#126 - TODO Test Logseq 〖Task〗' -> '(bw)Is#126 - TODO Test Logseq [[Task]]'
``` 
and then `task list` includes
```
23 1min   MyLogseqGraph                                    (bw)Is#126 - TODO Test Logseq &open;&open;Task&close;&close;        1
```

Reply:
Can you see what `task` invocation bugwarrior-pull is running in that case?

Reply:
Looks like the problem could be within the `taskw` library. I ran a simple test
```python
from taskw import TaskWarrior
w = TaskWarrior()
w.task_add("Task added using [[taskw]]")
```
```python
{'id': 24,
 'description': 'Task added using &open;&open;taskw&close;&close;',
 'entry': '20240628T102423Z',
 'modified': '20240628T102423Z',
 'status': 'pending',
 'uuid': '9f1544f9-e66e-41ca-b65e-79c328bac19c',
 'urgency': 0}
```

Reply:
specifically this appears to be the cause https://github.com/ralphbean/taskw/blob/3256c92bed4bb3d53269c843e8201e9ac65780cc/taskw/utils.py#L27

`taskw` is escaping the `[`, `]` and `"` characters before invoking `task` command line.
```python
[b'task', b'rc.verbose=new-uuid', b'rc.json.array=TRUE', b'rc.confirmation=no', b'rc.dependency.confirmation=no', b'rc.recurrence.confirmation=no', b'9f1544f9-e66e-41ca-b65e-79c328bac19c', b'modify', b'description:"Task added using &open;&open;taskw&close;&close; + (does) {this} &dquot;quote&dquot; work?"', b'entry:"20240628T102423Z"', b'modified:"20240628T110925Z"', b'status:"pending"']
```

Reply:
Thanks! I've updated the taskwarrior bug.

Reply:
Created https://github.com/ralphbean/taskw/issues/172 to track the actual underlying bug in taskw. I'm thinking if we documented at the beginning of this section that this is a workaround for https://github.com/ralphbean/taskw/issues/172 to give a little more context to these options that would resolve this thread.

Reply:
Updated with reference to the `taskw` issue in the code and the service doc.

Reply:
Thanks. Maybe I'm reading too much into it but I'm curious if there was intent behind the phrase "in part" where you say "This capability is *in part* to workaround...". I presume you'd be fine with deprecating and eventually removing this feature once the taskw bug is fixed?

Reply:
Specifically for the double braces `[[` and `]]` I would consider leaving the capability in place, to allow alternate styling of the page references.

Reply:
It would be great to update the comment to say that `taskw` does this encoding, not Taskwarrior.

Reply:
Already updated in commit https://github.com/GothenburgBitFactory/bugwarrior/pull/1052/commits/7a94d57f663b89b8445360e234772c0c1f75deb3

Reply:
Oh, sorry, the GitHub UI confused me

Reply:
> Specifically for the double braces `[[` and `]]` I would consider leaving the capability in place, to allow alternate styling of the page references.

Does Logseq itself have a feature to change this styling? To me this would seem like a vanity feature which doesn't provide enough value to justify its maintenance and documentation if it were not for the taskw bug it serves to work around.

*Note: Resolving this difference of opinion is not blocking this PR at this point and can be postponed until the upstream bug is fixed.*